processing/src/main/java/org/apache/druid/frame/processor/SuperSorter.java

```diff
@@ -0,0 +1,868 @@
+      final List<OutputChannel> channels = new ArrayList<>(partitions.size());
+
+      for (int partitionNum = 0; partitionNum < partitions.size(); partitionNum++) {
+        channels.add(outputChannelFactory.openNilChannel(partitionNum));
+      }
+
+      // OK to use wrap, not wrapReadOnly, because nil channels are already read-only.
+      allDone.set(OutputChannels.wrap(channels));
+    } else if (totalMergingLevels != UNKNOWN_LEVEL
+               && outputsReadyByLevel.containsKey(totalMergingLevels - 1)
+               && outputsReadyByLevel.get(totalMergingLevels - 1).size() == getOutputPartitions().size()) {
+      // We're done!!
+      try {
+        // OK to use wrap, not wrapReadOnly, because all channels in this list are already read-only.
+        allDone.set(OutputChannels.wrap(outputChannels));
+      }
+      catch (Throwable e) {
+        allDone.setException(e);
+      }
+    }
+  }
+
+  @GuardedBy("runWorkersLock")
+  private boolean runNextBatcher()
+  {
+    if (batcherIsRunning || inputChannelsToRead.isEmpty()) {
+      return false;
+    } else {
+      batcherIsRunning = true;
+
+      runWorker(
+          new FrameChannelBatcher(inputChannels, maxChannelsPerProcessor),
+          result -> {
+            final List<Frame> batch = result.lhs;
+            final IntSet keepReading = result.rhs;
+
+            synchronized (runWorkersLock) {
+              inputBuffer.addAll(batch);
+              inputFramesReadSoFar += batch.size();
+              inputChannelsToRead = keepReading;
+
+              if (inputChannelsToRead.isEmpty()) {
+                inputChannels.forEach(ReadableFrameChannel::doneReading);
+                setTotalInputFrames(inputFramesReadSoFar);
+                runWorkersIfPossible();
+              } else if (inputBuffer.size() >= maxChannelsPerProcessor) {
+                runWorkersIfPossible();
+              }
+
+              batcherIsRunning = false;
+            }
+          }
+      );
+
+      return true;
+    }
+  }
+
+  /**
+   * Level zero mergers read batches of frames from the "inputBuffer". These frames are individually sorted, but there
+   * is no ordering between the frames. Their output is a sorted sequence of frames.
+   */
+  @GuardedBy("runWorkersLock")
+  private boolean runNextLevelZeroMerger()
+  {
+    if (inputBuffer.isEmpty() || (inputBuffer.size() < maxChannelsPerProcessor && !allInputRead())) {
+      return false;
+    }
+
+    final List<ReadableFrameChannel> in = new ArrayList<>();
+
+    while (in.size() < maxChannelsPerProcessor) {
+      final Frame frame = inputBuffer.poll();
+
+      if (frame == null) {
+        break;
+      }
+
+      in.add(singleReadableFrameChannel(new FrameWithPartition(frame, FrameWithPartition.NO_PARTITION)));
+    }
+
+    runMerger(0, levelZeroMergersRunSoFar++, in, null);
+    return true;
+  }
+
+  @GuardedBy("runWorkersLock")
+  private boolean runNextMiddleMerger()
```

Comment:
Very cool implementation, great use of concurrency.

The first impression is that this on class does too much: it will be very hard to test thoroughly because of the amount of things that happen at once. From the school of hard knocks, I've found it best to split up a sort into its component pieces, test each to death, then combine them into the overall sort.

Here, for example, the actual mid-level merger could be a separate class (or function) so that it can be tested in isolation (two empty inputs, n empty inputs, all inputs the same value, lopsided inputs, already-sorted inputs chunks of inputs are in ascending order, but don't overlap, etc.)

The complex code to orchestrate "threads" seems similar to that used elsewhere: that is multiple copies to review and test. Can there be a single runner that handles generic tasks? We test the heck out of that once, and trust it in other uses.

In each "concurrent aware" operation, there is a simple sequential operation at its core. separating the two (very different) concerns will allow easier testing of each. That is, for this function, each merge is single-threaded, but coordinates with other workers at the boundaries.

And so on.

Reply:
The actual merger is indeed a separate class: FrameChannelMerger (single-threaded frame processor that does an N-way merge). The logic in SuperSorter is mainly concerned with creating FrameChannelMergers and connecting them to each other such that the right thing happens. Definitely open to suggestions about how it could be made simpler, or which parts might be ripe for extraction elsewhere.

Reply:
Ah! Glad to hear that the "guts" are in a separate bit we can test well.

I wouldn't block a commit for restructuring; we can tackle it later as needed. I find that, for things like this, it's hard to know a better structure until we have to change something: then struggles with tests (or understanding) suggest the joints at which the code should be carved.